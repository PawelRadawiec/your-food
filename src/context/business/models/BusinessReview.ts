export interface BusinessReview {
  id: string;
  rating: number;
  user: ReviewUser;
  text: string;
  time_created: string;
  url: string;
}

export interface ReviewUser {
  id: string;
  profile_url: string;
  image_url: string;
  name: string;
}
