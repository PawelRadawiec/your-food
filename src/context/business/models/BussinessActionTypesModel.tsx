export enum BussinessActionTypes {
  SET_RESULTS = 'SET_RESULTS',
  SET_RESULTS_MAP = 'SET_RESULTS_MAP',
  SET_LOADING = 'SET_LOADING',
  SET_REVIEWS_PENDING = 'SET_REVIEWS_PENDING',
  SET_BUSINESS = 'SET_BUSINESS',
  SET_REVIEWS = 'SET_REVIEWS',
  SELECTED_PENDING = 'SELECTED_PENDING',
  REQUEST_ERROR = 'REQUEST_ERROR',
  SET_PARAMS = 'SET_PARAMS'
}
