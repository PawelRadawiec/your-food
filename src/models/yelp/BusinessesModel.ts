import { Category } from './CategoryModel';
import { Coordinates } from './CoordinatesModel';
import { Location } from './LocationModel';

export interface BusinessesModel {
  rating: number;
  price: string;
  phone: string;
  id: string;
  alias: string;
  is_closed: boolean;
  categories: Category[];
  review_count: number;
  name: string;
  url: string;
  coordinates: Coordinates[];
  image_url: string;
  location: Location;
  distance: number;
  transactions: string[];
}
