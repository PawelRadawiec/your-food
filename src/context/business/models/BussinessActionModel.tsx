import { BussinessActionTypes } from "./BussinessActionTypesModel";

export interface BussinessAction {
  type: BussinessActionTypes;
  payload: any;
}
