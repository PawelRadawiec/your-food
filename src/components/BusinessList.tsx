import React, { useContext, useEffect, useState } from 'react';
import { FlatList, Image, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { BallIndicator } from 'react-native-indicators';
import { of } from 'rxjs';
import BussinessContext from '../context/business/BusinesseContext';
import { BusinessesModel } from '../models/yelp/BusinessesModel';
import StarsList from './StarsList';

const BusinessList = ({ data, navigation }: { data: BusinessesModel[]; navigation: any }) => {
  const {
    state: { selectedPending, params, loading, resultsMap },
    actions,
  } = useContext(BussinessContext);
  const [resultKeys, setResultsKyes] = useState<string[]>([]);
  useEffect(() => {
    setResultsKyes([...resultsMap.keys()]);
  }, [resultsMap]);
  return (
    <FlatList
      data={resultKeys}
      keyExtractor={(item) => item}
      renderItem={({ item }) => {
        return (
          <View>
            <Text style={styles.type}>{item}</Text>
            <FlatList
              showsHorizontalScrollIndicator={false}
              horizontal={true}
              onEndReached={() => {
                if (!loading) {
                  // todo - on end params
                  // actions.search(params);
                }
              }}
              data={resultsMap.get(item)}
              renderItem={({ item }) => {
                return (
                  <View style={styles.business}>
                    {selectedPending.id === item.id && selectedPending.pending ? (
                      <BallIndicator size={20} />
                    ) : (
                      <TouchableOpacity
                        onPress={() => {
                          if (!selectedPending.pending) {
                            actions.getById(item.id, navigation);
                          }
                        }}
                      >
                        <Text style={styles.title}>{item.name}</Text>
                      </TouchableOpacity>
                    )}
                    <View style={styles.information}>
                      <View>
                        <Text style={styles.city}>{item.location?.city}</Text>
                      </View>
                      <View>
                        <Text style={styles.city}>Reviews {item.review_count}</Text>
                      </View>
                      <View>
                        {/* @todo - fix horizontal*/}
                        {/* <StarsList rating={Number(item.rating)} /> */}
                      </View>
                    </View>
                    <Image
                        style={styles.image}
                        source={
                          item.image_url
                            ? {
                                uri: item.image_url,
                              }
                            : require('../images/no_image.jpeg')
                        }
                      />
                  </View>
                );
              }}
              keyExtractor={(item) => item.id}
            />
          </View>
        );
      }}
    />
  );
};

const styles = StyleSheet.create({
  business: {
    marginHorizontal: 10,
  },
  image: {
    width: 350,
    height: 200,
  },
  information: {
    flexDirection: 'column',
  },
  title: {
    fontWeight: 'bold',
    fontSize: 18,
  },
  city: {
    fontWeight: 'bold',
    fontSize: 14,
  },
  type: {
    fontWeight: 'bold',
    fontSize: 20,
    textTransform: 'capitalize',
    marginLeft: 10
  }
});

export default BusinessList;
