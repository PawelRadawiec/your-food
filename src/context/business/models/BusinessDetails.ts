import { Category } from '../../../models/yelp/CategoryModel';
import { Coordinates } from '../../../models/yelp/CoordinatesModel';
import { Location } from '../../../models/yelp/LocationModel';

export interface BusinessDetails {
  id: string;
  alias: string;
  name: string;
  image_url: string;
  is_claimed: string;
  is_closed: string;
  url: string;
  phone: string;
  display_phone: string;
  review_count: string;
  categories: Category[];
  rating: string;
  location: Location;
  coordinates: Coordinates;
  photos: string[];
  price: string;
  hours: BusinessHour[];
  transactions: [];
  special_hours: [];
}

export interface BusinessHour {
  hours_type: string;
  is_open_now: boolean;
  open: BusinessDayDetails[];
}

export interface BusinessDayDetails {
  day: number;
  end: 'string';
  is_overnight: boolean;
  start: string;
}
