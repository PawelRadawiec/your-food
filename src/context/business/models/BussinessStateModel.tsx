import { BusinessesModel } from '../../../models/yelp/BusinessesModel';

export interface BussinessState {
  results: BusinessesModel[];
  loading: boolean;
}

export const defaulBusinesstState: BussinessState = {
  results: [],
  loading: false,
};
