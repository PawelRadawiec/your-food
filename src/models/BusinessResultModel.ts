import { BussinessSearchParams } from './api/BusinessSearchParams';
import { BusinessesModel } from './yelp/BusinessesModel';

export interface BusinessResult {
  params: BussinessSearchParams;
  businesses: BusinessesModel[];
}
