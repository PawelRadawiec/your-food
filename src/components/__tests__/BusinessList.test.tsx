import React from 'react';
import Enzyme, { mount, ReactWrapper, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import BusinessList from '../BusinessList';
import { BusinessResult } from '../../models/BusinessResultModel';
import BussinessContext, { BusinessActions } from '../../context/business/BusinesseContext';
import { FlatList, TouchableOpacity } from 'react-native';
import { BusinessesModel } from '../../models/yelp/BusinessesModel';

Enzyme.configure({ adapter: new Adapter() });

const originalConsoleError = console.error;
console.error = (message) => {
  if (message.startsWith('Warning:')) {
    return;
  }
  originalConsoleError(message);
};

describe('business list component unit tests', () => {
  let wrapper: ReactWrapper;
  let getByIdMock: any;
  let searchMock: any;
  let actions: BusinessActions;

  beforeEach(() => {
    getByIdMock = jest.fn();
    searchMock = jest.fn();
    const actionsPartial: Partial<BusinessActions> = {
      getById: getByIdMock,
      search: searchMock,
    };
    actions = actionsPartial as BusinessActions;
    wrapper = mount(
      <BussinessContext.Provider
        value={{
          state: getDefaultState(),
          actions: actions as BusinessActions,
        }}
      >
        <BusinessList navigation={() => {}} />
      </BussinessContext.Provider>
    );
  });

  it('should render', () => {
    const component = shallow(<BusinessList navigation={() => {}} />);
    expect(component.getElements()).toMatchSnapshot();
  });

  it('should contains resultKeys', () => {
    const component = wrapper.find(BusinessList);
    const resultKeys = component.find(FlatList).first().prop('data') as string[];
    expect(resultKeys?.length).toBeGreaterThan(0);
  });

  it('should render horizontal flat list', () => {
    const component = wrapper.find(BusinessList);
    const hotizontalList = component.findWhere((node) => node.props().testID === 'pizza').first();
    expect(hotizontalList?.prop('data')?.length).toBeGreaterThan(0);
    expect(hotizontalList?.prop('horizontal')).toBeTruthy();
  });

  it('should call getById', () => {
    const component = wrapper.find(BusinessList);
    const hotizontalList = component.findWhere((node) => node.props().testID === 'pizza').first();
    hotizontalList.find(TouchableOpacity).first().simulate('click');
    expect(getByIdMock.mock.calls.length).toBe(1);
  });

  it('should not call getById', () => {
    const state = getDefaultState();
    state.selectedPending = { pending: true, id: '' };
    wrapper = mount(
      <BussinessContext.Provider
        value={{
          state,
          actions: actions as BusinessActions,
        }}
      >
        <BusinessList navigation={() => {}} />
      </BussinessContext.Provider>
    );
    const component = wrapper.find(BusinessList);
    const hotizontalList = component.findWhere((node) => node.props().testID === 'pizza').first();
    hotizontalList.find(TouchableOpacity).first().simulate('click');
    expect(getByIdMock.mock.calls.length).toBe(0);
  });

  it('should call search on end reached', () => {
    const component = wrapper.find(BusinessList);
    const hotizontalList = component.findWhere((node) => node.props().testID === 'pizza').first();
    hotizontalList.prop('onEndReached')();
    expect(searchMock.mock.calls.length).toBe(1);
  });

  it('should not call search on end reached', () => {
    const state = getDefaultState();
    state.loading = true;
    wrapper = mount(
      <BussinessContext.Provider
        value={{
          state,
          actions: actions as BusinessActions,
        }}
      >
        <BusinessList navigation={() => {}} />
      </BussinessContext.Provider>
    );
    const component = wrapper.find(BusinessList);
    const hotizontalList = component.findWhere((node) => node.props().testID === 'pizza').first();
    hotizontalList.prop('onEndReached')();
    expect(searchMock.mock.calls.length).toBe(0);
  });
});

const getDefaultState = () => {
  const resultsMap = new Map<string, BusinessResult>();
  const business1: Partial<BusinessesModel> = {};
  const business2: Partial<BusinessesModel> = {};
  resultsMap.set('pizza', {
    params: {},
    businesses: [business1 as BusinessesModel, business2 as BusinessesModel],
  });
  return {
    reviews: [],
    loading: false,
    business: null,
    selectedPending: { pending: false, id: '' },
    reviewsPending: false,
    resultsMap: resultsMap,
  };
};
