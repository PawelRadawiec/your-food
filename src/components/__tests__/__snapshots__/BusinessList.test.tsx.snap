// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`business list component unit tests should render 1`] = `
Array [
  <FlatList
    contentContainerStyle={
      Object {
        "paddingBottom": 70,
      }
    }
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />,
]
`;
